freqtrade/exchange/__init__.py

```diff
@@ -48,390 +40,379 @@ def wrapper(*args, **kwargs):
-        raise DependencyException(
-            f'Insufficient funds to create limit sell order on market {pair}.'
-            f'Tried to sell amount {amount} at rate {rate} (total {rate*amount}).'
-            f'Message: {e}')
-    except ccxt.InvalidOrder as e:
-        raise DependencyException(
-            f'Could not create limit sell order on market {pair}.'
-            f'Tried to sell amount {amount} at rate {rate} (total {rate*amount}).'
-            f'Message: {e}')
-    except (ccxt.NetworkError, ccxt.ExchangeError) as e:
-        raise TemporaryError(
-            f'Could not place sell order due to {e.__class__.__name__}. Message: {e}')
-    except ccxt.BaseError as e:
-        raise OperationalException(e)
-
-
-@retrier
-def get_balance(currency: str) -> float:
-    if _CONF['dry_run']:
-        return 999.9
-
-    # ccxt exception is already handled by get_balances
-    balances = get_balances()
-    balance = balances.get(currency)
-    if balance is None:
-        raise TemporaryError(
-            f'Could not get {currency} balance due to malformed exchange response: {balances}')
-    return balance['free']
-
-
-@retrier
-def get_balances() -> dict:
-    if _CONF['dry_run']:
-        return {}
-
-    try:
-        balances = _API.fetch_balance()
-        # Remove additional info from ccxt results
-        balances.pop("info", None)
-        balances.pop("free", None)
-        balances.pop("total", None)
-        balances.pop("used", None)
-
-        return balances
-    except (ccxt.NetworkError, ccxt.ExchangeError) as e:
-        raise TemporaryError(
-            f'Could not get balance due to {e.__class__.__name__}. Message: {e}')
-    except ccxt.BaseError as e:
-        raise OperationalException(e)
-
-
-@retrier
-def get_tickers() -> Dict:
-    try:
-        return _API.fetch_tickers()
-    except ccxt.NotSupported as e:
-        raise OperationalException(
-            f'Exchange {_API.name} does not support fetching tickers in batch.'
-            f'Message: {e}')
-    except (ccxt.NetworkError, ccxt.ExchangeError) as e:
-        raise TemporaryError(
-            f'Could not load tickers due to {e.__class__.__name__}. Message: {e}')
-    except ccxt.BaseError as e:
-        raise OperationalException(e)
-
-
-@retrier
-def get_ticker(pair: str, refresh: Optional[bool] = True) -> dict:
-    global _CACHED_TICKER
-    if refresh or pair not in _CACHED_TICKER.keys():
+class Exchange(object):
+
+    # Current selected exchange
+    _API: ccxt.Exchange = None
+    _CONF: Dict = {}
+    _CACHED_TICKER: Dict[str, Any] = {}
+
+    # Holds all open sell orders for dry_run
+    _DRY_RUN_OPEN_ORDERS: Dict[str, Any] = {}
+
+    def __init__(self, config: dict) -> None:
+        """
+        Initializes this module with the given config,
+        it does basic validation whether the specified
+        exchange and pairs are valid.
+        :param config: config to use
+        :return: None
+        """
+        self._API
```

Comment:
statement has no effect

Reply:
err ... yes, apparently
wonder that it passed both flake8 and mypy checks ...

Reply:
It is PEP8 conform and uses no typehints, so no complains :-) AFAIK pylint or pycharm should give you a warning

Reply:
pylint warns, yes, but it shows 93 warnings/problems for exchange/__init__.py alone ... so i don't really pay too much attention to it...